### PR TITLE
Count the attention matrix products of an F.scaled_dot_product_attention block

### PR DESCRIPTION
**Pull request by raimbekovm**
## Summary

Adds `count_scaled_dot_product_attention`, a counting rule for self-attention blocks built from `F.scaled_dot_product_attention` (the qkv-Linear + SDPA + output-Linear pattern used by `timm` and DINOv2/DINOv3-style vision transformers). The functional call has no owning module, so `add_hooks` never reaches it, and the block's own MACs come back short by the whole O(N²) attention term while the qkv/output `nn.Linear` children are still counted correctly.

The new rule is not auto-registered — there is no fixed torch type to key it to, since this attention shape is always a caller-defined class — so it is meant to be wired up per class through the existing `custom_ops` parameter, the same mechanism `README.md` already documents for third-party modules. README gains a short section showing that usage.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds support for counting the attention matrix products performed by `F.scaled_dot_product_attention` in custom self-attention blocks.

### 📊 Key Changes
- Adds `count_scaled_dot_product_attention` to `thop/vision/basic_hooks.py`.
- Counts query-key and attention-value matrix products while relying on existing hooks for linear projections.
- Documents registration and usage with a ViT-style attention example in `README.md`.

### 🎯 Purpose & Impact
- Enables more complete MAC estimates for attention implementations using PyTorch’s functional scaled dot-product API.
- Provides a reusable custom operation rule for blocks that cannot be hooked at the functional call level.
- Improves profiling accuracy for self-attention models, while intentionally omitting softmax operations due to unavailable generic block metadata.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/thop` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/thop/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/thop` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. This is an automated message, and an engineer will assist with any follow-up questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The counting formula and documented custom-op registration match the shown self-attention block, including the expected MAC and parameter totals. The README should disclose the PyTorch version requirement for the functional attention API.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean PR. The new custom hook correctly adds the two self-attention matrix products while preserving the existing child `nn.Linear` counts, and the README example’s expected MAC and parameter totals are consistent with the implementation.

**Comment by glenn-jocher** [comment]
Closing — this is the one I most wanted to keep, but the helper is silently wrong for the blocks that motivate it.

`count_scaled_dot_product_attention` reads `x[0].shape[-2:]` as `(tokens, dim)`. That holds for a ViT block taking `(B, N, C)`, and the README example is arithmetically correct (I checked: 113,246,208 + 37,748,736 + 50,331,648 = 201,326,592 ✓).

But Ultralytics' attention blocks take `(B, C, H, W)`. Running this branch's helper on one:

```
helper on a (B,C,H,W) attention block : 4,194,304
actual attention matmuls              : 134,217,728
error                                 : -96.88%
```

It read `shape[-2:] = (32, 32)` — H and W — as tokens and dim.

So for `Attention` (`block.py:1309`) and `AAttn` (`block.py:1695`), which is where our transformer FLOPs actually go missing, registering this rule reports ~3% of the truth with no warning. And the blocks it *would* fit — `TransformerEncoderLayer`, `AIFI`, `DeformableTransformerDecoderLayer` — all use `nn.MultiheadAttention` and are already covered by `count_multihead_attention` (#157).

A shipped formula that guesses the caller's tensor layout is worse than no formula: uncounted work is visible via `report_missing`, a wrong count is not. The existing `custom_ops` mechanism — already documented in the README, and accepting any callable since #155 with call args since #156 — lets a user write the two correct lines for their own layout.

If you want to pick this back up, the version worth having is on the **Ultralytics** side: `custom_ops` entries for `Attention`/`AAttn`/`MSDeformAttn` that read `B, C, H, W` from those blocks' own signatures. #167, which I kept, is what makes that safe — without it a quadratic custom rule gets extrapolated affinely by the `stride=` path and lands 36% low.
